src/main/java/io/gravitee/policy/json2xml/JsonToXmlTransformationPolicy.java

```diff
@@ -40,62 +41,81 @@
  */
 public class JsonToXmlTransformationPolicy extends JsonToXmlTransformationPolicyV3 implements Policy {
 
+    private static final String MESSAGE_INVALID_PAYLOAD_FAILURE_KEY = "JSON_INVALID_MESSAGE_PAYLOAD";
+
     public JsonToXmlTransformationPolicy(final JsonToXmlTransformationPolicyConfiguration configuration) {
         super(configuration);
     }
 
+    private static void applyHeaders(HttpHeaders headers, Buffer xmlBuffer) {
+        headers.set(HttpHeaderNames.CONTENT_TYPE, CONTENT_TYPE);
+        headers.set(HttpHeaderNames.CONTENT_LENGTH, Integer.toString(xmlBuffer.length()));
+    }
+
     @Override
     public String id() {
         return "json-xml";
     }
 
     @Override
     public Completable onRequest(final HttpExecutionContext ctx) {
-        return ctx
-            .request()
-            .onBody(
-                bodyUpstream ->
-                    bodyUpstream
-                        .flatMap(buffer -> transformToXml(buffer, CharsetHelper.extractCharset(ctx.request().headers())))
-                        .doOnSuccess(xmlBuffer -> applyHeaders(ctx.request().headers(), xmlBuffer))
-            )
-            .onErrorResumeNext(throwable -> interruptWithBadRequest(ctx, throwable));
+        return ctx.request().onBody(body -> transformBodyToXml(ctx, body, ctx.request().headers(), HttpStatusCode.BAD_REQUEST_400));
```

Comment:
maybe I lose something, but on Message style, we update the content headers with the type and size 
Here, we do not to it manually so the headers are not updated.

By the way, read my next comment about a little refactoring for that.